src/nnet3/nnet-discriminative-training.h

```diff
@@ -0,0 +1,146 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef KALDI_NNET3_NNET_DISCRIMINATIVE_TRAINING_H_
+#define KALDI_NNET3_NNET_DISCRIMINATIVE_TRAINING_H_
+
+#include "nnet3/nnet-example.h"
+#include "nnet3/nnet-computation.h"
+#include "nnet3/nnet-compute.h"
+#include "nnet3/nnet-optimize.h"
+#include "nnet3/nnet-discriminative-example.h"
+#include "nnet3/nnet-training.h"
+#include "nnet3/discriminative-training.h"
+
+namespace kaldi {
+namespace nnet3 {
+
+struct NnetDiscriminativeTrainingOptions {
+  NnetTrainerOptions nnet_config;
+  discriminative::DiscriminativeTrainingOptions discriminative_training_config;
+  discriminative::DiscriminativeTrainingStatsOptions discriminative_training_stats_config;
+
+  bool apply_deriv_weights;
+
+  NnetDiscriminativeTrainingOptions(): apply_deriv_weights(true) { }
+
+  void Register(OptionsItf *opts) {
+    nnet_config.Register(opts);
+    discriminative_training_config.Register(opts);
+    discriminative_training_stats_config.Register(opts);
+    opts->Register("apply-deriv-weights", &apply_deriv_weights,
+                   "If true, apply the per-frame derivative weights stored with "
+                   "the example (you'll normally want to leave this as true.");
+  }
+};
+
+// This struct is used in multiple nnet training classes for keeping
+// track of objective function values.
+// Also see struct AccuracyInfo, in nnet-diagnostics.h.
+struct DiscriminativeObjectiveFunctionInfo {
+  int32 current_phase;
+
+  double tot_aux_objf;
+  discriminative::DiscriminativeTrainingStats stats;
+  
+  double tot_aux_objf_this_phase;
+  discriminative::DiscriminativeTrainingStats stats_this_phase;
+
+  DiscriminativeObjectiveFunctionInfo():
+      current_phase(0), tot_aux_objf(0.0),
+      tot_aux_objf_this_phase(0.0) { }
+
+  // This function updates the stats and, if the phase has just changed,
+  // prints a message indicating progress.  The phase equals
+  // minibatch_counter / minibatches_per_phase.  Its only function is to
+  // control how frequently we print logging messages.
+  void UpdateStats(const std::string &output_name,
+                   const std::string &criterion,
+                   int32 minibatches_per_phase,
+                   int32 minibatch_counter,
+                   BaseFloat this_minibatch_weight,
+                   BaseFloat this_minibatch_tot_objf,
+                   BaseFloat this_minibatch_tot_aux_objf = 0.0);
+  
+  void UpdateStats(const std::string &output_name,
```

Comment:
OK, move it inside.
And you shouldn't overload functions like this- you need to update the
names to something meaningful.
Dan


On Tue, Mar 8, 2016 at 6:52 PM, Vimal Manohar <notifications@github.com>
wrote:

> In src/nnet3/nnet-discriminative-training.h
> <https://github.com/kaldi-asr/kaldi/pull/563#discussion_r55450416>:
>
> > +      current_phase(0), tot_aux_objf(0.0),
> > +      tot_aux_objf_this_phase(0.0) { }
> > +
> > +  // This function updates the stats and, if the phase has just changed,
> > +  // prints a message indicating progress.  The phase equals
> > +  // minibatch_counter / minibatches_per_phase.  Its only function is to
> > +  // control how frequently we print logging messages.
> > +  void UpdateStats(const std::string &output_name,
> > +                   const std::string &criterion,
> > +                   int32 minibatches_per_phase,
> > +                   int32 minibatch_counter,
> > +                   BaseFloat this_minibatch_weight,
> > +                   BaseFloat this_minibatch_tot_objf,
> > +                   BaseFloat this_minibatch_tot_aux_objf = 0.0);
> > +
> > +  void UpdateStats(const std::string &output_name,
>
> I can move the xent objf inside the same structure.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/563/files#r55450416>.
>